### ISSUE TITLE
pluto hangs in connect() when configured with Opportunistic IPsec over TCP

### ISSUE THREAD

**Issue by ueno** [libreswan/libreswan#1156]
With the following configuration, pluto seems to hang in connect():

```console
$ cat /etc/ipsec.d/oa.conf
conn private
        type=tunnel
        authby=null
        leftid=%null
        rightid=%null
        left=%defaultroute
        right=%opportunisticgroup
        negotiationshunt=hold
        failureshunt=drop
        ikev2=insist
        auto=route
        enable-tcp=yes
        tcp-remoteport=4500

$ ping 10.0.0.3
...
20 packets transmitted, 0 received, 100% packet loss, time 19460ms
```

It works if `enable-tcp` is commented. Here is a backtrace:
```console
#0  0x00007fce6d17548b in connect () from /lib64/libc.so.6
#1  0x0000559b40089edd in connect_to_tcp_endpoint (local_dev=0x559b41c04388, 
    remote_endpoint=..., logger=0x559b41c094b8)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/iface_tcp.c:487
#2  0x0000559b400456ec in initiate_v2_IKE_SA_INIT_request (c=0x559b41c0b5c8, 
    predecessor=0x0, policy=25904095257, try=1, inception=0x7ffc115a01b0, 
    sec_label=..., background=true, logger=0x7ffc115a4e00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/ikev2_ike_sa_init.c:545
#3  0x0000559b4009ab0b in initiate_ondemand_body (b=b@entry=0x7ffc115a0b00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/initiate.c:1006
#4  0x0000559b4009bc2c in initiate_ondemand (
    packet=packet@entry=0x7ffc115a2d20, by_acquire=by_acquire@entry=true, 
    background=background@entry=true, sec_label=..., 
    logger=logger@entry=0x7ffc115a4e00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/initiate.c:1027
#5  0x0000559b40075751 in netlink_acquire (logger=0x7ffc115a4e00, 
    n=0x7ffc115a0d10)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/kernel_xfrm.c:1874
#6  netlink_get (fd=fd@entry=12, logger=logger@entry=0x7ffc115a4e00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/kernel_xfrm.c:2146
#7  0x0000559b40075a73 in netlink_process_msg (fd=12, logger=0x7ffc115a4e00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/kernel_xfrm.c:2175
#8  0x0000559b400652f9 in kernel_process_msg_cb (fd=12, 
    arg=0x559b401ce240 <xfrm_kernel_ops>, logger=0x7ffc115a4e00)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/kernel.c:2482
#9  0x0000559b400aa60c in fd_read_listener_event_handler (fd=<optimized out>, 
    events=<optimized out>, arg=<optimized out>)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/server.c:795
#10 0x00007fce6d2deb2c in event_process_active_single_queue ()
   from /lib64/libevent_core-2.1.so.7
#11 0x00007fce6d2e080f in event_base_loop () from /lib64/libevent_core-2.1.so.7
#12 0x0000559b400acd83 in run_server (
    conffile=0x559b41b3bc18 "/etc/ipsec.conf", logger=0x7ffc115a5090)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/server.c:1070
#13 0x0000559b40041ad9 in main (argc=<optimized out>, argv=<optimized out>)
    at /usr/src/debug/libreswan-4.11-1.fc38.x86_64/programs/pluto/plutomain.c:1821
```
Here is the output of `lsof -nP -iTCP`, on the connecting side:
```console
pluto     11949            root   14u  IPv4  35767      0t0  TCP 10.0.0.2:40848->10.0.0.3:4500 (SYN_SENT)
pluto     11949            root   17u  IPv4  35678      0t0  TCP 10.0.0.2:4500 (LISTEN)
pluto     11949            root   20u  IPv4  35681      0t0  TCP 127.0.0.1:4500 (LISTEN)
```
on the remote side:
```console
pluto     11513            root   17u  IPv4  36066      0t0  TCP 10.0.0.3:4500 (LISTEN)
pluto     11513            root   20u  IPv4  36069      0t0  TCP 127.0.0.1:4500 (LISTEN)
```

Tested with libreswan-4.11-1.fc38 on Fedora 38 Server spin, which doesn't enable firewalld.

**Comment by cagney** [comment]
Is this a duplicate of #553

**Comment by ueno** [comment]
> Is this a duplicate of https://github.com/libreswan/libreswan/issues/553

I'm actually not sure, as this blocks for a long time (if not forever). Is there anything that could prevent TCP handshake completing?

**Comment by cagney** [comment]
It should eventually timeout.  All connect(2) as is:
```
       ETIMEDOUT
              Timeout  while  attempting connection.  The server may be too busy to accept new connections.  Note that for IP sockets the timeout may be very long when syncookies are enabled on
              the server.
```

**Comment by ueno** [comment]
@cagney I indeed see `#1: TCP: connect(14) failed: Connection timed out (errno 110)` but note that this only happens with Opportunistic IPsec setup. With the same `listen-tcp=yes` / `enable-tcp=yes`, host-to-host configuration works without hang:
```
conn mytunnel
	type=tunnel
	authby=null
	leftid=%null
	rightid=%null
	left=10.0.186.114
	right=10.0.186.102
	negotiationshunt=hold
	failureshunt=drop
	ikev2=insist
	auto=route
	enable-tcp=yes
	tcp-remoteport=4500
```

Do you know anything peculiar to Opportunistic IPsec that could cause `connect` to block?

**Comment by cagney** [comment]
> Do you know anything peculiar to Opportunistic IPsec that could cause connect to block?

Not really.  It's the same connect call.  I do wonder, though, about the kernel policy and the need to let through TCP.  Perhaps compare the kernel policy for various scenarios such as:
- OE triggered by packet (add, route, ping)
- OE started manually (add, up)
- same for on-demand
- same for permenant

**Comment by ueno** [comment]
OK, after a bit more testing, the above host-to-host [configuration](https://github.com/libreswan/libreswan/issues/1156#issuecomment-2187931456) works only after manual add/up. The configuration has `auto=route` but the packet based trigger only works with UDP.

Before add, a trap policy is installed:
```console
src 10.0.187.15/32 dst 10.0.186.153/32 
	dir out priority 1753281 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 0 mode transport
```
which is removed after `ipsec auto --add mytunnel`, and after up, the following policies are installed:
```console
src 10.0.187.15/32 dst 10.0.186.153/32 
	dir out priority 1753281 ptype main 
	tmpl src 10.0.187.15 dst 10.0.186.153
		proto esp reqid 16389 mode tunnel
src 10.0.186.153/32 dst 10.0.187.15/32 
	dir fwd priority 1753281 ptype main 
	tmpl src 10.0.186.153 dst 10.0.187.15
		proto esp reqid 16389 mode tunnel
src 10.0.186.153/32 dst 10.0.187.15/32 
	dir in priority 1753281 ptype main 
	tmpl src 10.0.186.153 dst 10.0.187.15
		proto esp reqid 16389 mode tunnel
```

**Comment by cagney** [comment]
> The configuration has `auto=route` but the packet based trigger only works with UDP.

`auto=route` is equivalent to
```
ipsec add ...
ipsec route ...
```

> 
> Before add, a trap policy is installed:
> 
> ```
> src 10.0.187.15/32 dst 10.0.186.153/32 
> 	dir out priority 1753281 ptype main 
> 	tmpl src 0.0.0.0 dst 0.0.0.0
> 		proto esp reqid 0 mode transport
> ```
> 
> which is removed after `ipsec auto --add mytunnel`, and after up, the following policies are installed:

Right.  `ipsec add ...` is equivalent to:
```
ipsec delete ... # removes existing policy and connection
ipsec add ...
```

> My guess is that, to exchange data for the initial IKE messages with TCP, it first needs to complete a TCP handshake, which requires a SYN-ACK packet needs to be sent back, though there is only one policy for out direction in the on-demand case.

Right, until that happens, the connect() blocks.  But if it doesn't happen it should time out.  The acquire code does roughly the following (assuming the acquire policy is installed):

- kernel installs a trap policy for what just happened
- libreswan replaces trap policy with block policy
- libreswan calls connect()

To your point, could the connect() be blocked by the kernel because its outgoing packet is stuck in the block policy. What does the TCP case look like before/after the acquire?  Oh and what happens if the OE selector is for a specific port (I suspect this will work because the trap won't cover the ESPINTCP connection) and when `negotiationpolicy=pass`.

**Comment by ueno** [comment]
Thanks for the hints; I'll look into it. FWIW, here is a test case that can reproduce the issue with kvm: [basic-pluto-14-ondemand-tcp.tar.gz](https://github.com/user-attachments/files/16048415/basic-pluto-14-ondemand-tcp.tar.gz)

**Comment by ueno** [comment]
Sorry, I forgot to add a comment here.

>  * kernel installs a trap policy for what just happened
>  * libreswan replaces trap policy with block policy
>  * libreswan calls connect()

In my testing, these are not happening in the order; at the entry to connect(), the kernel trap policy is still there and no block policy is installed.

However, if I move the call to `kernel_ops->poke_ipsec_policy_hole` in `connect_to_tcp_endpoint` before connect() as in #1746, the blocking problem disappears (in both OE and non-OE cases). Questions are:
- Is this the right approach?
- If yes, should it be conditionalized in the ondemand case only; it's not necessary in the manual `ipsec up` case, because a fwd policy is installed

**Comment by cagney** [comment]
> Is this the right approach?

I think your right.  Your evidence is pretty compelling.  For the `ipsec up` case I don't think it does harm.   And I suspect:

- `ipsec up`
- peer crashes
- local end tries to recover which means putting in an on-demand trap like for `ipsec route`

would fail.

**Comment by cagney** [comment]
fixed by @ueno
